accepted/future-releases/language-versioning/package-config-file-v2.md

```diff
@@ -0,0 +1,134 @@
+
+Instead of breaking the existing file users, we will instead introduce a *new* file containing the information that we intended to store in the `.packages` file. This fill will use a (more) structured format which allows extending it with further information in the future. We will still generate the `.packages` file for now, allowing users of the file to continue working until they can migrate to using the new file.
+
+## New File Contents
+
+The file will contain the information that was planned for the `.packages` file:
+
+- A *default package* specified by its package name.
+- A list of *packages*, each with:
+  - A *package name*.
+  - A *path* or *location URI*.
+  - A *language version* (major/minor version number)
+
+On top of that, we also add extra metadata which allows us to version the format and attach information about when and how the file was created.
+
+We use JSON for the format. JSON has parsers available for most languages (not all Dart tools are written in Dart) and it is standardized and efficiently parsable.  JSON is *not* easy to write by hand (but doable), and it does not allow comments, but this is a machine generated file that will hardly ever be edited by hand.
+
+## New File Format
+
+The new file will be called `.packages.json` and stored in the same directory as the `.packages` file. The format is JSON text describing is a single JSON object.
+
+That object has, at least, the following properties:
+
+- `apiVersion`: A single integer marking the version of this format, currently `2` (counting `.packages` as the first). Tools should refuse to work on files with a larger version number than the ones they are aware of. Incremental changes will not require increasing the version, so new versions are not expected any time soon, but code should still prepare for it.
+
+- `defaultPackage`: A string containing a package name.
+- `packages`: A list of packages, each with the following properties:
+  - `name`: A string containing a package name.
+  - either `path`: A string containing a relative or absolute path using `/` as separator. Every character other than `/` is used verbatim. If the path starts with a `/` or with a DOS/Windows drive letter (a single letter followed by `:/`), then it is considered absolute. If not, the path is relative to the directory of the `.packages.json` file. 
+  - or `uri`: A URI reference. This is resolved relative to the (likely `file:`) URI referencing the `.package.json` file. If the URI reference has no scheme, or if it has a `file:` scheme, it is recommended to use the `path` instead.
+  - `languageVersion`: Optional. A string containing a language version (major.minor).
+
+Tools should ignore properties that they don't recognize. This will allow us to add more properties in the future without increasing the `apiVersion` number, but not to remove properties or change the meaning of existing properties. 
```

Comment:
Also, tools that read in and then output a `.packages.json` file should preserve all data in the file even if the tool itself doesn't use it. This way the tool doesn't accidentally strip out extensions used by other tools.

Reply:
I deliberately avoided that. Do you think it is a good idea? I'm not sure I *want* tools to put arbitrary data into the file. If we do that, we'll likely see `analysis_options.yaml` getting moved into this file (I mean, why not, the analyzer will be reading the file anyway?). 
Is this what we *want*? An arbitrary registry of metadata for dart files in the related subdir?

We would also have to specify what it means, exactly, to "preserve data". Should Pub retain a package entry for a package it doesn't know about (any more)? What if that entry contains a property unknown to Pub? Most likely we will say that existing package entries *can* be removed by any tool generating package entries, but if the same package still exists, then any extra properties that Pub doesn't overwrite should be retained.

Should Pub retain extra information in the `generator` entry, or is that entirely its own?

Reply:
> I deliberately avoided that. Do you think it is a good idea?

Yes, it's essentially required. If we allow tools to store their own extensions in the file (and I think we do want to do that), then we need to ensure other tools roundtrip those files must preserve those extensions.

In practice, I don't think it will be much of an issue. This file isn't intended to be the source of truth for data. Most tools will either only read it or only write it. Tools that do generate this file from scratch are not obliged to maintain extensions put in some previous incarnation of the file generated by a different tool. It's basically on the user to ensure that two tools which generate the same file somehow coordinate with each other.

Reply:
Ah, so you are not talking about pub, but explicitly about tools which modify the Pub-created file for their own purposes (adding data that only they understand).

If Pub does not retain extra data, but creates it from scratch, then other tools which add something on top are at Pub's mercy. Since Pub may be called at any time by your editor if you touch `pubspec.yaml` for any reason, you can't rely on that extra data anyway (unless those tools are called *by* Pub every time it generates the file.

If we want to preserve extra data, I think even Pub should do so, but that risks accumulating cruft over time. At least users can always delete the file and start over.

I'm not sure we *do* want to allow tools to store their own extensions in the file. It is not the Windows Registry for Dart tools, the file has a purpose. Unless we want to have a registry for  Dart tools. Then retaining the unknown content should be

Reply:
> Ah, so you are not talking about pub, but explicitly about tools which modify the Pub-created file for their own purposes (adding data that only they understand).

Pub isn't a good example because it never reads any file that it also writes. Pub only reads pubspec.yaml files and it only writes pubspec.lock files. Thus it never roundtrips a file or has to worry about preserving extensions.

But tools like stagehand that have commands for modifying a pubspec *do* read and write the same file and have to make sure to preserve data.